### PR TITLE
Implemented buckets

### PR DESCRIPTION
**Pull request by TomCaserta**
Added items for buckets and placement. I am unable to figure out how to get the block face for the bukkit event on line 46 in ItemBucket.java. However the rest of the implementation works.

### CODE REVIEW

**Comment by TomCaserta** [comment]
Wait what, I literally just tested this exact code before creating the pull request... Hang on

File: src/main/java/net/glowstone/block/blocktype/BlockLava.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package net.glowstone.block.blocktype;
+
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockLava.java]: Whitespace

File: src/main/java/net/glowstone/block/blocktype/BlockWater.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+public class BlockWater extends BlockType {
+
+    @Override
+    public void placeBlock(GlowPlayer player, GlowBlockState state, BlockFace face, ItemStack holding, Vector clickedLoc) {
+        state.setType(getMaterial());
+        state.setRawData((byte)8);
+    }
+
+}
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockWater.java]: All files need a single newline character at the end

File: src/main/java/net/glowstone/block/itemtype/ItemBucket.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
+package net.glowstone.block.itemtype;
+
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java]: Whitespace

File: src/main/java/net/glowstone/block/itemtype/ItemBucket.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    public ItemBucket () {
+        super();
+        this.setMaxStackSize(16);
+    }
+
+    @Override
+    public void rightClickAir(GlowPlayer player, ItemStack holding) {
+        Iterator<Block> itr = new BlockIterator(player, 5);
+        Block target = null;
+        while (itr.hasNext()) {
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java:36]: This section of code gets the "last target"?

File: src/main/java/net/glowstone/block/itemtype/ItemBucket.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                replaceWith = Material.LAVA_BUCKET;
+            } else {
+                return;
+            }
+
+            if (target.getState().getRawData() != 8) {
+                return;
+            }
+
+            // Todo: Are bucket fill events supposed to fire even if its not water or lava?
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java]: No

File: src/main/java/net/glowstone/block/itemtype/ItemFilledBucket.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import net.glowstone.block.GlowBlockState;
+import net.glowstone.entity.GlowPlayer;
+import org.bukkit.GameMode;
+import org.bukkit.Material;
+import org.bukkit.block.BlockFace;
+import org.bukkit.event.player.PlayerBucketEmptyEvent;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.util.Vector;
+
+abstract public class ItemFilledBucket extends ItemPlaceAs {
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemFilledBucket.java]: `public abstract`

File: src/main/java/net/glowstone/block/itemtype/ItemBucket.java (hunk 4 above)
Comment (TomCaserta) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java:36]: First non air block target.

Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java:36]: There must be a better way...

File: src/main/java/net/glowstone/block/itemtype/ItemBucket.java (hunk 5 above)
Comment (EntityReborn) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java]: What about milk?

Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java]: I think the question was more or less "should they fire for things other than liquids?" ie: slamming a bucket into a door.

Comment (TomCaserta) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java]: Yeah, wasn't sure if for some reason someone needed the ability for their plugin to check if you tried filling a bucket with a random block.

Comment (EntityReborn) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java]: Maybe we can extend an interface and check instanceof?

Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java]: Yea, they can check to see if you tried to fill a bucket with dirt when that is more possible (if it's not already) via PlayerInteractEvent

Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java]: > Maybe we can extend an interface and check instanceof?

The client doesn't know how to deal with this.

**Comment by Tonodus** [comment]
Instead of using BlockWater and BlockLava, you might want to use BlockLiquid for both.

File: src/main/java/net/glowstone/block/itemtype/ItemFilledBucket.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        super(placeAs);
+    }
+
+    @Override
+    public void rightClickBlock(GlowPlayer player, GlowBlock against, BlockFace face, ItemStack holding, Vector clickedLoc) {
+        GlowBlock target = against.getRelative(face);
+        GlowBlockState newState = target.getState();
+
+        // only allow placement inside tall-grass, air, or liquid
+        if (against.getType() == Material.LONG_GRASS || against.getType() == Material.FIRE) {
```
Comment (EntityReborn) [inline src/main/java/net/glowstone/block/itemtype/ItemFilledBucket.java]: Hardcoding things like this is a bit of a pain. Could we have a BlockType.isReplaceable() or something?

**Comment by TomCaserta** [comment]
Just found a case where empty buckets are not working correctly, going to sort it now.

File: src/main/java/net/glowstone/block/blocktype/BlockFire.java [hunk 8]
```diff
@@ -11,3 +11,5 @@ public boolean canOverride(GlowBlock block, BlockFace face, ItemStack holding) {
         return true;
     }
 }
+
+
```
Comment (EntityReborn) [inline src/main/java/net/glowstone/block/blocktype/BlockFire.java]: This isn't necessary.

File: src/main/java/net/glowstone/block/blocktype/BlockLiquid.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+    public abstract boolean isCollectible(GlowBlockState block);
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Overrides
+
+    @Override
+    public void placeBlock(GlowPlayer player, GlowBlockState state, BlockFace face, ItemStack holding, Vector clickedLoc) {
+        state.setType(getMaterial());
+        state.setRawData((byte)8);
```
Comment (EntityReborn) [inline src/main/java/net/glowstone/block/blocktype/BlockLiquid.java:46]: Might be a good idea to place a comment by magic numbers like this.

File: src/main/java/net/glowstone/block/blocktype/BlockFire.java (hunk 8 above)
Comment (TomCaserta) [inline src/main/java/net/glowstone/block/blocktype/BlockFire.java]: Yeah currently fixing it up, was introduced when I fixed a merge conflict. Would you recommend I open a new PR without a ton of commits?

Comment (EntityReborn) [inline src/main/java/net/glowstone/block/blocktype/BlockFire.java]: There's been lots of comments here. Keep this, maybe squash everything?

**Comment by TomCaserta** [comment]
Finally got that all squashed. Hmm I could in-fact just use the same class for water and lava bucket...

**Comment by EntityReborn** [comment]
Keep Lava, Water and MilkBukkits in seperate classes, and have them inherit a common class with the functionality, if you don't already.

**Comment by EntityReborn** [comment]
I think the raw function of MilkBuckets (a bucket full of milk) would merit subclassing ItemBucketFillable. Any other functionality would be amended to that class.

**Comment by Tonodus** [comment]
Well, milk buckets aren't really buckets. It doesn't make that much sense ti implement them now (no potion, no cows), anyway it couldnlt extend any liquid class.

**Comment by TomCaserta** [comment]
Accidentally deleted my comment above instead of closing what I believed was the edit window... But yeah I agree with Tonodus, I feel any implementation of a milk bucket would have to be changed down the road anyway because we have no entity interact methods yet. I mean I could do dummy implementation but it seems pointless at the moment.

**Comment by EntityReborn** [comment]
Valid points. Carry on :)

File: src/main/java/net/glowstone/block/ItemTable.java [hunk 10]
```diff
@@ -1,9 +1,7 @@
 package net.glowstone.block;
 
 import net.glowstone.block.blocktype.*;
-import net.glowstone.block.itemtype.ItemPlaceAs;
-import net.glowstone.block.itemtype.ItemSign;
-import net.glowstone.block.itemtype.ItemType;
+import net.glowstone.block.itemtype.*;
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/ItemTable.java:4]: Not sure on our standard for this. @SpaceManiac  ?

File: src/main/java/net/glowstone/block/blocktype/BlockFire.java [hunk 11]
```diff
@@ -11,3 +11,4 @@ public boolean canOverride(GlowBlock block, BlockFace face, ItemStack holding) {
         return true;
     }
 }
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockFire.java]: :(

File: src/main/java/net/glowstone/block/blocktype/BlockLava.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    public BlockLava () {
+        this.setBucketType(Material.LAVA_BUCKET);
+    }
+
+    public boolean isCollectible (GlowBlockState target) {
+        return target.getType() == Material.STATIONARY_LAVA || (target.getType() == Material.LAVA && target.getRawData() == 8);
+    }
+
+}
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockLava.java]: :(

File: src/main/java/net/glowstone/block/blocktype/BlockLiquid.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+    /**
+     * Get the bucket type to replace the empty bucket when the liquid has been collected
+     * @return The associated bucket types material
+     */
+    public Material getBucketType () {
+        return this.bucketType;
+    }
+
+    public abstract boolean isCollectible(GlowBlockState block);
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockLiquid.java:36]: Docs?

File: src/main/java/net/glowstone/block/blocktype/BlockWater.java [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+    public BlockWater () {
+        this.setBucketType(Material.WATER_BUCKET);
+    }
+
+    public boolean isCollectible (GlowBlockState target) {
+        return target.getType() == Material.STATIONARY_WATER || (target.getType() == Material.WATER && target.getRawData() == 8);
+    }
+
+}
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/blocktype/BlockWater.java]: :(

File: src/main/java/net/glowstone/block/itemtype/ItemBucket.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+                    holding.setAmount(holding.getAmount() - 1);
+                    player.getInventory().addItem(new ItemStack(replaceWith));
+                }
+            }
+
+            target.setType(Material.AIR);
+        }
+    }
+}
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java]: :(

File: src/main/java/net/glowstone/block/itemtype/ItemLavaBucket.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+
+public class ItemLavaBucket extends ItemFilledBucket {
+
+    public ItemLavaBucket () {
+        super(Material.LAVA);
+        this.setMaxStackSize(1);
+    }
+
+}
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemLavaBucket.java]: :(

File: src/main/java/net/glowstone/block/itemtype/ItemWaterBucket.java [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+
+public class ItemWaterBucket extends ItemFilledBucket {
+
+    public ItemWaterBucket () {
+        super(Material.WATER);
+        this.setMaxStackSize(1);
+    }
+
+}
+
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemWaterBucket.java]: :(

File: src/main/java/net/glowstone/block/ItemTable.java (hunk 10 above)
Comment (EntityReborn) [inline src/main/java/net/glowstone/block/ItemTable.java:4]: I think we were thinking anything 4 imports and up is ok to asterisk.

Comment (SpaceManiac) [inline src/main/java/net/glowstone/block/ItemTable.java:4]: Yes, something along those lines. While the namespace pollution is real, I much prefer it over repeating `import net.glowstone.block.itemtype.X` 5 million times in a case like this where it isn't just a design problem.

Comment (turt2live) [inline src/main/java/net/glowstone/block/ItemTable.java:4]: Alright. So the 4th import makes it a *?

**Comment by turt2live** [comment]
Sorry - bot fail :(

File: src/main/java/net/glowstone/block/itemtype/ItemBucket.java [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+            if (targetBlockType != null && targetBlockType instanceof BlockLiquid) {
+                if (((BlockLiquid)targetBlockType).isCollectible((GlowBlockState) target.getState())) {
+                    validTarget = true;
+                    break;
+                }
+            }
+        }
+
+        if (target != null && validTarget == true) {
+            // TODO: Do some math to get the actual block face.
```
Comment (turt2live) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java]: This needs to be fixed before this can be pulled

Comment (TomCaserta) [inline src/main/java/net/glowstone/block/itemtype/ItemBucket.java]: Aight fair enough, well, if someone else is able to submit a PR to this PR as I have no idea what the maths for that is as mentioned in the first post.

**Comment by TomCaserta** [comment]
All done! Figured I could just get the face from knowing what block is directly before in the line of sight and using that to determine block face.

Edit: of course I forget to remove the glowserver import I was using to debug.

**Comment by turt2live** [comment]
@TomCaserta Your PR does not currently merge. Please resolve the issue as soon as possible! Thanks.

**Comment by turt2live** [comment]
@TomCaserta Your PR has not been mergable for at least 14.7887297282523 days! Please update your PR as soon as possible, thanks!

`PR_BOT_CODE:2` (you can ignore this)

**Comment by turt2live** [comment]
@TomCaserta Your PR has not been mergable for at least 3.51442965089583 days! Please update your PR as soon as possible, thanks!

`PR_BOT_CODE:2` (you can ignore this)

**Comment by dequis** [comment]
Merged branch: https://github.com/dequis/Glowstone/compare/pr-236

A single but non-trivial conflict in EventFactory due to a refactoring in d3ed79ea7d284df1d2cd1945bf53d5652962a34f, fix was to change:

EventFactory.onPlayerBucketFill(...) -> EventFactory.callEvent(new PlayerBucketFillEvent(...))
EventFactory.onPlayerBucketEmpty(...) -> EventFactory.callEvent(new PlayerBucketBucketEmpty(...))

Also checkstyle fixes.

**Comment by turt2live** [comment]
@TomCaserta Due to recent changes on master, this PR is no longer mergable. Please rebase as soon as possible!

**Comment by dequis** [comment]
Updated branch, no new conflicts.

**Comment by turt2live** [comment]
@TomCaserta Your PR is now not mergable due to recent changes. Please update your PR as soon as you can, thanks!

**Comment by turt2live** [comment]
The supplied branch (https://github.com/dequis/Glowstone/compare/pr-236) is ready for pulling as of this comment. The known issues are as follows and are to be fixed at a later date:
- Buckets in adventure mode are functional by default
- Buckets in spectator mode are functional by default

Thanks for your contribution @TomCaserta and thank you @dequis for updating this as needed ;)

**Comment by SpaceManiac** [comment]
Merged in 3334d1d0917d64818d1b9feeba0a8ab6dc299975.

**Pull request closed without merging**